### PR TITLE
Chapter 11

### PR DESCRIPTION
**Pull request by Phamduc8797**
C9 Implement advanced login

### CODE REVIEW

**Comment by Phamduc8797** [comment]
ready

File: app/assets/javascripts/account_activations.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (huuphuc93) [inline app/assets/javascripts/account_activations.coffee]: remove this file

File: app/assets/stylesheets/account_activations.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the AccountActivations controller here.
```
Comment (huuphuc93) [inline app/assets/stylesheets/account_activations.scss]: remove this file

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -4,9 +4,16 @@ def new; end
   def create
     user = User.find_by email: params[:session][:email].downcase
     if user&.authenticate params[:session][:password]
-      log_in user
-      params[:session][:remember_me] == "1" ? remember(user) : forget(user)
-      redirect_back_or user
+      if user.activated?
+        log_in user
+        params[:session][:remember_me] == '1' ? remember(user) : forget(user)
```
Comment (huuphuc93) [inline app/controllers/sessions_controller.rb]: use " instead of '

File: app/mailers/user_mailer.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class UserMailer < ApplicationMailer
+  def account_activation user
+    @user = user
+    mail to: user.email, subject: t ".subject"
+  end
+
+  def password_reset
+    @greeting = "Hi"
```
Comment (huuphuc93) [inline app/mailers/user_mailer.rb]: I18n

File: app/models/user.rb [hunk 5]
```diff
@@ -27,12 +28,32 @@ def remember
...
+  def activate
+    update_attributes activated: true
+    update_attributes activated_at: Time.zone.now
+  end
+
+  def send_activation_email
+    UserMailer.account_activation(self).deliver_now
+  end
+
+  private
```
Comment (huuphuc93) [inline app/models/user.rb]: add empty line after private

File: app/models/user.rb [hunk 6]
```diff
@@ -27,12 +28,32 @@ def remember
...
+    def downcase_email
+      self.email = email.downcase
```
Comment (huuphuc93) [inline app/models/user.rb]: email.downcase!

File: app/models/user.rb [hunk 7]
```diff
@@ -27,12 +28,32 @@ def remember
...
+    end
+
+    def create_activation_digest
+      self.activation_token  = User.new_token
+      self.activation_digest = User.digest(activation_token)
```
Comment (huuphuc93) [inline app/models/user.rb]: remove ()

File: app/views/user_mailer/password_reset.html.erb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+<h1>User#password_reset</h1>
+
+<p>
+  <%= @greeting %>, find me in app/views/user_mailer/password_reset.html.erb
```
Comment (huuphuc93) [inline app/views/user_mailer/password_reset.html.erb]: I18n

File: test/controllers/account_activations_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (huuphuc93) [inline test/controllers/account_activations_controller_test.rb]: remove this file

File: test/mailers/previews/user_mailer_preview.rb [hunk 10]
```diff
@@ -0,0 +1,16 @@
+# Preview all emails at http://localhost:3000/rails/mailers/user_mailer
```
Comment (huuphuc93) [inline test/mailers/previews/user_mailer_preview.rb]: remove this file

File: test/mailers/user_mailer_test.rb [hunk 11]
```diff
@@ -0,0 +1,20 @@
+require 'test_helper'
```
Comment (huuphuc93) [inline test/mailers/user_mailer_test.rb]: remove this file

**Comment by huuphuc93** [comment]
commented

File: app/models/user.rb [hunk 12]
```diff
@@ -27,12 +28,33 @@ def remember
...
+  end
+
+  def send_activation_email
+    UserMailer.account_activation(self).deliver_now
+  end
+
+  private 
+
+    def downcase_email
+      self.email = email.downcase!
```
Comment (huuphuc93) [inline app/models/user.rb]: self.email = email.downcase! => email.downcase!

**Comment by huuphuc93** [comment]
commented
comment nhỏ

**Comment by Phamduc8797** [comment]
ready

**Comment by huuphuc93** [comment]
merged
this (pr)  can be (merged)  now
CC: [To:2076564] Nguyen Huy Hung C (93)
CC: [To:1546746] Md. Mahmudur Rahman (Mahmud)
CC: [To:1818111] Nguyen Thi Huong (94)

**Pull request merged**